plugins/inputs/nsd/nsd.go

```diff
@@ -0,0 +1,184 @@
+package nsd
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"net"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/filter"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type runner func(cmdName string, Timeout internal.Duration, UseSudo bool, Server string, ConfigFile string) (*bytes.Buffer, error)
+
+// NSD is used to store configuration values
+type NSD struct {
+	Binary     string
+	Timeout    internal.Duration
+	UseSudo    bool
+	Server     string
+	ConfigFile string
+
+	filter filter.Filter
+	run    runner
+}
+
+var defaultBinary = "/usr/sbin/nsd-control"
+var defaultTimeout = internal.Duration{Duration: time.Second}
+
+var sampleConfig = `
+  ## Address of server to connect to, read from nsd conf default, optionally ':port'
+  ## Will lookup IP if given a hostname
+  server = "127.0.0.1:8953"
+
+  ## If running as a restricted user you can prepend sudo for additional access:
+  # use_sudo = false
+
+  ## The default location of the nsd-control binary can be overridden with:
+  # binary = "/usr/sbin/nsd-control"
+
+  ## The default location of the nsd config file can be overridden with:
+  # config_file = "/etc/nsd/nsd.conf"
+
+  ## The default timeout of 1s can be overridden with:
+  # timeout = "1s"
+`
+
+// Description displays what this plugin is about
+func (s *NSD) Description() string {
+	return "A plugin to collect stats from the NSD authoritative DNS name server"
+}
+
+// SampleConfig displays configuration instructions
+func (s *NSD) SampleConfig() string {
+	return sampleConfig
+}
+
+// Shell out to nsd_stat and return the output
+func nsdRunner(cmdName string, Timeout internal.Duration, UseSudo bool, Server string, ConfigFile string) (*bytes.Buffer, error) {
+	cmdArgs := []string{"stats_noreset"}
+
+	if Server != "" {
+		host, port, err := net.SplitHostPort(Server)
+		if err != nil { // No port was specified
+			host = Server
+			port = ""
+		}
+
+		resolver := net.Resolver{}
+		ctx, lookUpCancel := context.WithTimeout(context.Background(), Timeout.Duration)
+		defer lookUpCancel()
+		serverIps, err := resolver.LookupIPAddr(ctx, host)
+		if err != nil {
+			return nil, fmt.Errorf("error looking up ip for server: %s: %s", Server, err)
+		}
+		if len(serverIps) == 0 {
+			return nil, fmt.Errorf("error no ip for server: %s: %s", Server, err)
+		}
+		server := serverIps[0].IP.String()
```

Comment:
I understand that nsd-control's -s option requires an IP address, not a hostname, so we need to resolve the name here.  I'm a little concerned that this ignores secondary IPs.  Ordinary, connecting through Dial() will try each IP in order until it succeeds (https://golang.org/pkg/net/#Dial).  It may be unlikely that we can't connect on the first IP and we can on a secondary, but if this is the case and we never try the secondary IPs, telegraf will fail silently and it will be difficult for a user to debug.

I wonder if it would be worth it to remove the ability to resolve the server hostname and require the user to provide an IP, since this is what nsd-control's -s option requires.

Reply:
Agree. In fact I add this because the `unbound` input plugin have this behavior.